### ISSUE TITLE
Field filter from nested query

### ISSUE THREAD

**Issue by biji** [metabase/metabase#6449]
I have created custom SQL Wheat Field filter and add the question to dashboard. I can add category filter and choose the field
And then I create another question based on previous question this is pie chart. And I add the question to the dashboard but I can not use variable from previous question for filtering

Thank you

**Comment by wvengen** [comment]
It would be very useful to propagate variables from the originating question to the derived question.

**Comment by eyalfir** [comment]
Same issue.
The Field Filter options for custom SQL questions is a life saver in term of flexibility to power users (=knows their SQL). However, that power user cannot democratise this by giving hist non-SQL-savvy colleagues the ability to ask derived questions. So in my case, most of the team cannot realistically benefit from custom SQL queries.

**Comment by brycemelvin** [comment]
I am trying to create a cumulative sum view of a filterable custom sql query but cannot pass the variables through on the cumulative sum view via the dashboard. This is a must have for us, unless someone knows another way to do a cumulative sum view in SQL.

**Comment by flamber** [comment]
Related to (maybe duplicate of) #4367

**Comment by mlm603** [comment]
similar issue to some of those in this thread, when we have a SQL question with a required field filter variable, we want to be able to build a Simple Question from it and filter the variable in the Simple Question. 

Running into some issues in v1.35

My SQL Question looks like this, with a field filter variable called `Created Date` that is a date type
![Screen Shot 2020-08-19 at 6 23 39 PM](https://user-images.githubusercontent.com/13624950/90695806-29343a00-e249-11ea-9165-91ce984fbe4f.png)

Then when I build a Simple Question off of that SQL Question, I see this `template_variable` filter option (which I assume is supposed to be connected to my `Created Date` variable?)
![Screen Shot 2020-08-19 at 6 25 00 PM](https://user-images.githubusercontent.com/13624950/90695950-6dbfd580-e249-11ea-94cc-fa219997ad00.png)

But then that `template_variable` filter only allows text inputs
![Screen Shot 2020-08-19 at 6 25 21 PM](https://user-images.githubusercontent.com/13624950/90696052-9b0c8380-e249-11ea-9e88-ba86e6990db0.png)

**Comment by pedrokost** [comment]
I am using a hacky workaround for this:

The question which includes the sub-query must re-define the variables from the sub-query. 

For example:

```
WITH subquery_args AS ( SELECT 1 WHERE {{user}} IS NULL),
all_devices AS {{#111}}
SELECT ....
```

Inside the CTE subquery_args I use the {{user}} variable which is also used inside the subquery {{#111}}. The subquery_args is formed in a way that it does not affect the query results in any way.

This is obviously a hack, which could break in the future (I hope it won't break until there is a proper solution). 

@ngnclht1102

**Comment by mayank924** [comment]
it throws an error! 
![image](https://user-images.githubusercontent.com/64311431/110512070-e4131f00-812a-11eb-8c76-e99b3ae04b6d.png)

**Comment by neewy** [comment]
This would be very helpful, upvoted 🙏🏻

**Comment by etoulas** [comment]
> I am using a hacky workaround for this:
> 
> The question which includes the sub-query must re-define the variables from the sub-query.
> 
> For example:
> 
> ```
> WITH subquery_args AS ( SELECT 1 WHERE {{user}} IS NULL),
> all_devices AS {{#111}}
> SELECT ....
> ```
> 
> Inside the CTE subquery_args I use the {{user}} variable which is also used inside the subquery `{{#111}}`. The subquery_args is formed in a way that it does not affect the query results in any way.
> 
> This is obviously a hack, which could break in the future (I hope it won't break until there is a proper solution).
> 
> @ngnclht1102

This works surprisingly well and it even works with Dashboards filters!

For multiple parameters just extend the pattern in the query.

```sql
WITH subquery_args AS ( SELECT 1 WHERE {{user}} IS NULL AND {{version}} IS NULL),
```

**Comment by etoulas** [comment]
@rouanes: I am using Postgres, which DB do you have?

**Comment by rouanes** [comment]
> @rouanes: I am using Postgres, which DB do you have?

Same actually. using Postgres.

**Comment by etoulas** [comment]
@rouanes: I'm afraid that I cannot help without your specific query. The FROM clause is not mandatory in Postgres, so there's another problem somewhere.

Here's a working example as a reference:

```sql
-- Original question
-- It will receive the variables from the derived question.
-- I saved it as quenstion_id = 123 and called it something like "power with arguments"
SELECT power({{base}}, {{exponent}}) AS "base raised to the power of exponent"
```

Base and exponent are defined as "Number" variables and are "required" with 2 as default values (to keep the showcase sane).

```sql
-- Derived question
-- Contains variables which get pushed down to the original question, simply by defining them.
-- The variables here have no other purpose in the derived questions.
WITH
  subquery_args AS (
    -- use the same variable names that you have in the original question.
    SELECT 1 WHERE {{base}} IS NULL AND {{exponent}} IS NULL
  )
, power_with_args AS {{#123}}  -- this is the original question from above!

SELECT *
FROM power_with_args
```

Make sure to define the variables with the same types as the variables in the original question.

**Comment by jatinForSigo** [comment]
I want to get the number of days between two particular dates. using metabase GUI, I'm able to get the difference in this `0 years 0 mons 521 days 4 hours 13 mins 42.0 secs` format but i can't convert this to number format.
Any help would be appreciated.
Thanks

**Comment by tjb9dc** [comment]
I can corroborate what @rouanes was saying, c @etoulas.

The "hacky" solution (which I think is an elegant way to trick Metabase to allow you to declare the variables 😄 ) works for the primitive filters like `Text`, `Number`, and `Date`, but it **does not** work for `Field Filter`. I get the same error:
```
ERROR: missing FROM-clause entry for table <table_your_first_field_filter_evaluates_against>
```

I tried it a bunch of times, including a verbatim swap back to `Text` which works.

Sad times, since field filters are so nice for the user

**Comment by dmarcolongo** [comment]
> I am using a hacky workaround for this:
> 
> The question which includes the sub-query must re-define the variables from the sub-query.
> 
> For example:
> 
> ```
> WITH subquery_args AS ( SELECT 1 WHERE {{user}} IS NULL),
> all_devices AS {{#111}}
> SELECT ....
> ```
> 
> Inside the CTE subquery_args I use the {{user}} variable which is also used inside the subquery `{{#111}}`. The subquery_args is formed in a way that it does not affect the query results in any way.
> 
> This is obviously a hack, which could break in the future (I hope it won't break until there is a proper solution).
> 
> @ngnclht1102

Great solution! I was able to make it even simpler by just inserting the variables into a comment line at the top (Metabase can`t tell the difference): 

` -- {{user}} {{#111}}`

**Comment by Netzvamp** [comment]
I've coded a solution, that pulls, merges and pushes questions from a local file repository. That way you can work around this issue in an elegant way.
https://discourse.metabase.com/t/mbs-manage-your-metabase-queries-with-local-files/21372 
https://github.com/Administerium/mbs

**Comment by Startouf** [comment]
Is this hack still working for recent metabse versions ? I'm on 0.43.1 and I'm getting errors 

> You'll need to pick a value for 'my_variable' before this query can run.

Even when using the comment of subquery_args trick with a variable of the same name

**Comment by Abdillah** [comment]
Providing manual dropdown options might be a temporary user experience patch for this issue.
Metabase could provide dropdown control that's not dependent on any column, instead data analysts should provide its value themselves. The audience then able to use this more intuitive filter rather than typing custom value..

**Comment by MarcSamD** [comment]
That's not completely the same case, but still find below my feedback about a closely related issue (**single question** with nested queries and Field Filters):

### First case:
This nested query works because the variable is a type "Text" (so not Field Filter)
![image](https://user-images.githubusercontent.com/70358310/204763677-e3f9916d-0511-42b3-88a3-802a4071d709.png)

### Second case:
If I try to convert the variable to a Field Filter, the query will not work because the variable directly refer to the column "PUBLIC.PEOPLE.SOURCE" instead of "table1.SOURCE"
![image](https://user-images.githubusercontent.com/70358310/204764404-fd75eb35-189f-49ae-89f6-3b78e6fad39f.png)

### Workaround:
Use another select clause in the where clause so that the Field Filter can directly be used.
![image](https://user-images.githubusercontent.com/70358310/204765805-b18ab782-55de-4777-815a-e1a80f8b4afe.png)

### Additional comment:
Obvisoulsy in this simple second case, it would have been much easier to just move the WHERE clause into the nested query. However you may need to keep the variable in the last WHERE clause when you are playing with more complex queries with several JOIN.

**Comment by BennoDev** [comment]
Is this ever happening? Definitely a big miss currently when trying to avoid duplication / work with base queries in large datasets.

**Comment by pedrokost** [comment]
Unfortunately, the workaround I had (see below) stopped working with Metabase 0.43. It works until 0.42.6. This is currently limiting my abilities to upgrade Metabase. Before upgrading I would have to modify several hundred saved questions and introduce a lot of duplicate SQL in each query.


> I am using a hacky workaround for this:
> 
> The question which includes the sub-query must re-define the variables from the sub-query.
> 
> For example:
> 
> ```
> WITH subquery_args AS ( SELECT 1 WHERE {{user}} IS NULL),
> all_devices AS {{#111}}
> SELECT ....
> ```
> 
> Inside the CTE subquery_args I use the {{user}} variable which is also used inside the subquery `{{#111}}`. The subquery_args is formed in a way that it does not affect the query results in any way.
> 
> This is obviously a hack, which could break in the future (I hope it won't break until there is a proper solution).
> 
> @ngnclht1102

**Comment by Telofy** [comment]
Same, except I already upgraded and just noticed that it's broken.

**Comment by Minishlink** [comment]
Stumbled upon this problem again today and I can confirm the workaround of re-using the variable in the parent query does not work in latest metabase version (v0.48.6)

**Comment by joeyvmason** [comment]
Looks like it's been 7 years since this has been addressed. For my use case this makes Metabase basically unusable. Is it likely this will be worked on?

**Comment by paoliniluis** [comment]
@farhanjo models are not meant to have filters, models are models, questions based on models can have filters

@joeyvmason we don't have this in the short time [roadmap](https://www.metabase.com/roadmap). Based on my 4 years in Metabase and also 3 more years implementing Metabase in several companies, I have never seen this issue as a blocker that would make Metabase unusable. Can you please explain your specific use case and why you say it's unusable?


tl;dr; if the problem is the duplication of work, we're addressing that separately on other issues

**Comment by wvengen** [comment]
This has not been a blocker issue for us in most cases. But we do have use-cases, including those that prevent us from using Metabase for those queries.

We mostly run into this issue when working with complex queries (usually involving several CTEs) that work with a large dataset, being very slow to work with as a whole. By putting a filter in the originating query, we can be very specific on how and where we filter, and make sure that the database planner has the information it needs to select the subset of the data to start with. If we _don't_ do this, most of the data is still being read, and then most of it is thrown away when filtering later on. This makes the queries usually time out.

Including more columns in the originating query could possibly help, then filtering on those in the derived questions (where we know the database planner can make a good link between the filter and underlying table columns). But then we need to move knowledge of how and where filtering happens, to the several derived questions, making the whole much harder to work with and maintain in case of changes.

**Comment by rodrigoGA** [comment]
This issue also affects models, and I believe it should be addressed there. For efficiency reasons, it’s sometimes necessary to create a persistent model, which is essentially a table.

The problem is that when using persistent models, the ability to apply field filters is lost because Metabase interprets them as subqueries. This is a major limitation since date and category filters, unless tied to a specific field, are too basic.

If this functionality cannot be implemented for subqueries, it should at least be available for persistent models. I don’t see a valid reason for this restriction, and it significantly limits the potential of models in Metabase.

**Comment by aphonsoar** [comment]
I would love to have this feature!

**Comment by aphonsoar** [comment]
Guys, this is extremely needed to avoid duplication of code! Please put in the roadmap, I beg you! 🙏

**Comment by kschutter** [comment]
This would be hugely helpful for me as well. I need to be able to work with complex variables in a sql question, and have the ability to  dig into the results from a dashboard. Currently, that would require me to have a "base" data question in the form of a table, then for every visualization, duplicate that base query, manually summarize it, then manually change the dashboard click-through to point back to my original base data. If I need to summarize by more than a few different columns, my base data starts to have too many variables to be usable (they would take up a large portion of the screen.)

**Comment by ranquild** [comment]
This can be done with parameters in snippets, starting from v57. You would need to extract your "base" query into a SQL snippet, define parameters there, and then create native queries that use this snippet. These queries can be added to a dashboard, and they would all inherit parameters from the base snippet.

**Comment by kschutter** [comment]
Snippet parameters is an ok workaround, but the 10k character limit, poor editing and source control, and being unable to use the parameterized-snippet for pivot tables makes this less than ideal.
